apps/application-system/api/src/app/modules/payment/payment.controller.ts

```diff
@@ -56,18 +61,64 @@ export class PaymentController {
   async createCharge(
     @CurrentUser() user: User,
     @Param('applicationId', new ParseUUIDPipe()) applicationId: string,
+    @Body() payload: CreateChargeInput,
   ): Promise<CreatePaymentResponseDto> {
     if (!isUuid(applicationId)) {
       throw new BadRequestException(`ApplicationId is on wrong format.`)
     }
+    const DISTRICT_COMMISSIONER_OF_REYKJAVIK = '6509142520'
+    const inputApplicationType = findItemType(payload.chargeItemCode)
+
+    // Finding application to confirm correct catalog & price
+    const thisApplication = await this.paymentService
+      .findApplicationById(applicationId, user.nationalId, inputApplicationType)
+      .catch((error) => {
+        throw new BadRequestException(
+          `Unable to find application with the ID ${applicationId} ` + error,
+        )
+      })
+
+    if (thisApplication.typeId.toString() !== inputApplicationType) {
+      throw new BadRequestException(
+        new Error(
+          'Mismatch between create charge input and application payment.',
+        ),
+      )
+    }
+
+    const allCatalogs =
+      payload.chargeItemCode.slice(0, 2) === 'AY'
+        ? await this.paymentAPI.getCatalogByPerformingOrg(
+            DISTRICT_COMMISSIONER_OF_REYKJAVIK,
+          )
+        : await this.paymentAPI.getCatalog()
+
+    // Sort through all catalogs to find the correct one.
+    const catalog = await this.paymentService
+      .searchCorrectCatalog(
+        payload.chargeItemCode,
+        JSON.stringify(allCatalogs.item),
```

Comment:
why stringify?

Reply:
Otherwise typescript throws this error in the service when trying to parse the catalog..
> Argument of type 'Item[]' is not assignable to parameter of type 'string'.ts(2345)
I could stringify inside the service, but I prefer to do it this way, it is cleaner.

Reply:
Maybe I am lacking some context, but why is the argument of type `string` in the first place? Looks like it is parsed straight away.

```typescript
const resultCatalog = JSON.parse(searchJSON)
```

Reply:
I have it in string format so it can easily be parsed. The service returns a [object, object] thus
I parse&for-loop the object so I can locate the correct catalog.
Is there a better way to go through [object, object] without json parsing?

Reply:
I think the problem is stringifying.. you definitely don't want that.. if you just want to search one of the items, and you are sure you are only getting a single item in allCatalogs, then you probably want something like:

```
const [ itemToSearchFor ] = allCatalogs.item

const catalog = await this.paymentService
      .searchCorrectCatalog(
        payload.chargeItemCode,
        itemToSearchFor,
```

Reply:
agree with @baering that the argument to the function should probably just be an item, if that's all you are ever searching for.